vowpalwabbit/core/src/learner.cc

```diff
@@ -312,12 +326,366 @@ bool ec_is_example_header(const example& ec, label_type_t label_type)
+  }
+  else if (_merge_f) { _merge_f(per_model_weighting, all_data, output_learner._learner_data.get()); }
+  else { THROW("learner " << _name << " does not support merging."); }
+}
+
+void learner::add(const VW::workspace& base_ws, const VW::workspace& delta_ws, const learner* base_l,
+    const learner* delta_l, VW::workspace& output_ws, learner* output_l)
+{
+  auto name = output_l->get_name();
+  assert(name == base_l->get_name());
+  assert(name == delta_l->get_name());
+  if (_add_with_all_f)
+  {
+    _add_with_all_f(base_ws, base_l->_learner_data.get(), delta_ws, delta_l->_learner_data.get(), output_ws,
+        output_l->_learner_data.get());
+  }
+  else if (_add_f) { _add_f(base_l->_learner_data.get(), delta_l->_learner_data.get(), output_l->_learner_data.get()); }
+  else { THROW("learner " << name << " does not support adding a delta."); }
+}
+
+void learner::subtract(const VW::workspace& ws1, const VW::workspace& ws2, const learner* l1, const learner* l2,
+    VW::workspace& output_ws, learner* output_l)
+{
+  auto name = output_l->get_name();
+  assert(name == l1->get_name());
+  assert(name == l2->get_name());
+  if (_subtract_with_all_f)
+  {
+    _subtract_with_all_f(
+        ws1, l1->_learner_data.get(), ws2, l2->_learner_data.get(), output_ws, output_l->_learner_data.get());
+  }
+  else if (_subtract_f)
+  {
+    _subtract_f(l1->_learner_data.get(), l2->_learner_data.get(), output_l->_learner_data.get());
+  }
+  else { THROW("learner " << name << " does not support subtraction to generate a delta."); }
+}
+
+std::unique_ptr<learner> learner::make_derived_learner()
+{
+  if (_derived_learner_created_already)
+  {
+    THROW("Cannot create more than one derived learner of base learner: " + _name);
+  }
+
+  // Copy this learner and assign this as the new learner's base
+  std::unique_ptr<learner> l(new learner(*this));
+  l->_base_learner = std::shared_ptr<learner>(this);
+
+  // We explicitly overwrite the copy of the base's finish_example functions.
+  // This is to allow us to determine if the current reduction implements finish
+  // and in what way.
+  l->_finish_example_f = nullptr;
+  l->_update_stats_f = nullptr;
+  l->_output_example_prediction_f = nullptr;
+  l->_print_update_f = nullptr;
+  l->_cleanup_example_f = nullptr;
+
+  // Don't propagate these functions
+  l->_multipredict_f = nullptr;
+  l->_save_load_f = nullptr;
+  l->_pre_save_load_f = nullptr;
+  l->_end_pass_f = nullptr;
+  l->_end_examples_f = nullptr;
+  l->_persist_metrics_f = nullptr;
+  l->_finisher_f = nullptr;
+
+  // Don't propagate any of the merge functions
+  l->_merge_f = nullptr;
+  l->_merge_with_all_f = nullptr;
+  l->_add_f = nullptr;
+  l->_add_with_all_f = nullptr;
+  l->_subtract_f = nullptr;
+  l->_subtract_with_all_f = nullptr;
+
+  _derived_learner_created_already = true;
+  return l;
+}
+
+learner& learner::safe_get_base_learner()
```

Comment:
Some docs here explaining what's meant by base_learner would make the expectation clear for the user

Reply:
I renamed the leaner immediately before the current learner as "previous". Now "base" refers specifically to the last learner of the entire reduction stack.